File: src/strategy/signal_engine.py
from dataclasses import dataclass


@dataclass
class Decision:
    symbol: str
    action: str
    quantity: int
    reason: str
    confidence: float


class SignalEngine:
    def __init__(self, buy_threshold: float = 0.65, sell_threshold: float = -0.65) -> None:
        self.buy_threshold = buy_threshold
        self.sell_threshold = sell_threshold

    def decide(self, symbol: str, sentiment: float, confidence: float, price: float) -> Decision:
        if confidence < 0.5 or price <= 0:
            return Decision(symbol, "HOLD", 0, "Low confidence or invalid price", confidence)
        if sentiment >= self.buy_threshold:
            return Decision(symbol, "BUY", 1, "Positive signal exceeded threshold", confidence)
        if sentiment <= self.sell_threshold:
            return Decision(symbol, "SELL", 1, "Negative signal exceeded threshold", confidence)
        return Decision(symbol, "HOLD", 0, "No threshold hit", confidence)

File: src_cpp/strategy/signal_engine.cpp
#include "strategy/signal_engine.hpp"

#include <algorithm>

SignalEngine::SignalEngine(double buy_threshold, double sell_threshold, double min_confidence, double max_trade_notional)
    : buy_threshold_(buy_threshold),
      sell_threshold_(sell_threshold),
      min_confidence_(min_confidence),
      max_trade_notional_(max_trade_notional) {}

int SignalEngine::position_size(double price) const {
    if (price <= 0.0) {
        return 0;
    }
    const int qty = static_cast<int>(max_trade_notional_ / price);
    return std::max(1, qty);
}

Decision SignalEngine::decide(const std::string& symbol, double sentiment, double confidence, double price) const {
    Decision d;
    d.symbol = symbol;
    d.confidence = confidence;

    if (price <= 0.0 || confidence < min_confidence_) {
        d.action = "HOLD";
        d.reason = "Low confidence or invalid price";
        d.quantity = 0;
        return d;
    }

    d.quantity = position_size(price);

    if (sentiment >= buy_threshold_) {
        d.action = "BUY";
        d.reason = "Positive signal exceeded threshold";
        return d;
    }

    if (sentiment <= sell_threshold_) {
        d.action = "SELL";
        d.reason = "Negative signal exceeded threshold";
        return d;
    }

    d.action = "HOLD";
    d.reason = "No threshold hit";
    d.quantity = 0;
    return d;
}
